File: src/matrix/matrix.go
package main

import (
    "fmt"
    "math/rand"
    "time"
)

// Function to generate a random matrix
func generateRandomMatrix(rows, cols int) [][]float64 {
    matrix := make([][]float64, rows)
    for i := 0; i < rows; i++ {
        matrix[i] = make([]float64, cols)
        for j := 0; j < cols; j++ {
            matrix[i][j] = rand.Float64()
        }
    }
    return matrix
}

// Function for matrix multiplication
func matrixMultiplication(a, b [][]float64) [][]float64 {
    rowsA := len(a)
    colsA := len(a[0])
    rowsB := len(b)
    colsB := len(b[0])

    if colsA != rowsB {
        panic("Incompatible matrices for multiplication")
    }

    result := make([][]float64, rowsA)
    for i := 0; i < rowsA; i++ {
        result[i] = make([]float64, colsB)
    }

    for i := 0; i < rowsA; i++ {
        for j := 0; j < colsB; j++ {
            for k := 0; k < colsA; k++ {
                result[i][j] += a[i][k] * b[k][j]
            }
        }
    }

    return result
}

func main() {
    start := time.Now()
    
    rand.Seed(time.Now().UnixNano())

    matrixSize := 1000 // Adjust the size of the matrix to increase memory usage
    fmt.Println("matrixSize:", matrixSize)

    matrixA := generateRandomMatrix(matrixSize, matrixSize)
    matrixB := generateRandomMatrix(matrixSize, matrixSize)
    matrixMultiplication(matrixA, matrixB)
    
    duration := time.Since(start)
    fmt.Println("Matrix Multiplication time:", duration)
}

File: src/matrix/matrix.rs
use std::time::Instant;

// Function to generate a random matrix
fn generate_random_matrix(rows: usize, cols: usize) -> Vec<Vec<f64>> {
    let mut matrix = Vec::with_capacity(rows);
    for _ in 0..rows {
        matrix.push((0..cols).map(|_| rand::random::<f64>()).collect());
    }
    matrix
}

// Function for matrix multiplication
fn matrix_multiplication(a: &Vec<Vec<f64>>, b: &Vec<Vec<f64>>) -> Vec<Vec<f64>> {
    let rows_a = a.len();
    let cols_a = a[0].len();
    let rows_b = b.len();
    let cols_b = b[0].len();

    if cols_a != rows_b {
        panic!("Incompatible matrices for multiplication");
    }

    let mut result = vec![vec![0.0; cols_b]; rows_a];

    for i in 0..rows_a {
        for j in 0..cols_b {
            for k in 0..cols_a {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }

    result
}

fn main() {
    let start_time = Instant::now();

    let matrix_size = 1000; // Adjust the size of the matrix to increase memory usage
    println!("matrixSize: {}", matrix_size);

    let matrix_a = generate_random_matrix(matrix_size, matrix_size);
    let matrix_b = generate_random_matrix(matrix_size, matrix_size);
    matrix_multiplication(&matrix_a, &matrix_b);

    let elapsed_time = start_time.elapsed();
    println!("Matrix Multiplication time: {:?}", elapsed_time);
}